### PR TITLE
Add debuginfod support

### LINKED ISSUE #1285: Add support for loading debug information from `debuginfod`
Fedora 35 release that is scheduled to the end of 2021 will use Debuginfod by default for storing debug information for the binaries:
- https://fedoraproject.org/wiki/Changes/DebuginfodByDefault

> With GNU Binutils 2.34 comes debuginfod support, which is the HTTP server catching our eye while the debuginfod server is distributed as part of the latest elfutils package. This isn't for a general purpose web server thankfully but is an HTTP server for distributing ELF/DWARF debugging information and source code. With debuginfod enabled, Binutils' `readelf` and `objdump` utilities can query the HTTP server(s) for debug files that cannot otherwise be found. Enabling this option requires building Binutils using `--with-debuginfod`. 

See https://developers.redhat.com/blog/2019/10/14/introducing-debuginfod-the-elfutils-debuginfo-server/ for more information.

- https://sourceware.org/elfutils/Debuginfod.html
- **Ubuntu Linux** https://ubuntu.com/server/docs/service-debuginfod
- **Debian Linux** https://wiki.debian.org/Debuginfod
- **Fedora Linux** https://fedoraproject.org/wiki/Debuginfod
- **Arch Linux** https://wiki.archlinux.org/title/Debuginfod

Depends on https://github.com/rizinorg/rizin/issues/1805

_Originally posted by @XVilka in https://github.com/rizinorg/rizin/issues/907#issuecomment-811727583_

### PR DESCRIPTION
**Pull request by b1llow**
Add rz_bin_dwarf_from_debuginfod

 

**Your checklist for this pull request**
- [ ] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [ ] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

...

**Test plan**

...

**Closing issues**

closes #1285

### CODE REVIEW

**Review by XVilka — COMMENTED** [review]
We also need to think about how to test this automatically without relying too much on external services. By running a debuginfod mock server of our own?

**Comment by b1llow** [comment]
> We also need to think about how to test this automatically without relying too much on external services. By running a debuginfod mock server of our own?

We can run a local private server when run CI.
https://sourceware.org/elfutils/Debuginfod.html

```
private servers
To run your own server, identify the directories containing raw build files, or RPM packages, or DEB packages, or any combination, plus any desired upstream federated servers. Launch debuginfod into the background, or use systemd.
% export DEBUGINFOD_URLS=https://foo.bar/   # list of federated upstream servers, if any
% debuginfod [-F $files] [-R $rpms] [-U $debs] &
Then, configure clients to talk to this server:
% export DEBUGINFOD_URLS=http://localhost:8002/
```

File: librz/bin/dwarf/dwarf.c [hunk 1]
```diff
@@ -185,6 +189,31 @@ RZ_API RZ_OWN RzBinDWARF *rz_bin_dwarf_search_debug_file_directory(
 	return NULL;
 }
 
+RZ_API RZ_OWN RzBinDWARF *rz_bin_dwarf_from_debuginfod(
+	RZ_BORROW RZ_NONNULL RzBinFile *bf,
+	RZ_BORROW RZ_NONNULL RzList /*<const char *>*/ *debuginfod_urls) {
```
Comment (XVilka) [inline librz/bin/dwarf/dwarf.c:194]: Add `rz_return_*()` for all `NONNULL`s.

**Review by XVilka — CHANGES_REQUESTED** [review]
Overall looks good. We need to make sure `debuginfod` is present on the CI for this test though, or maybe run it only on Linux systems, just like debugging tests.

**Review by XVilka — APPROVED** [review]
LGTM. @wargio @ret2libc, take a look, too, please.

**Comment by XVilka** [comment]
@imbillow you need to install it alson on Travis:
```
[XX] db/archos/linux-any/debuginfod "Debugging Information in debuginfod"

RZ_NOPLUGINS=1 /home/travis/build/rizinorg/rizin/install/bin/rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -eflirt.sigdb.load.system=false -eflirt.sigdb.load.home=false -N -Qc '& !debuginfod -vvv -F bins/elf/float_ex1/float_ex1_buildid bins/elf/float_ex1/float_ex1_split-dwarf.debug

sleep 10

e bin.dbginfo.debuginfod=true

e bin.dbginfo.debuginfod_urls=http://127.0.0.1:8002

o bins/elf/float_ex1/float_ex1_buildid

aaa

iS

afl

pdf @ dbg.fn1

qyn

' =

-- stdout

--- expected

+++ actual

@@ -39,71 +39,12 @@

 0x00001080    1 6            sym.imp.__cxa_finalize

 0x00001170    1 9            entry.init0

 0x000013b0    1 9            sym._fini

-0x00001180    5 170          dbg.fn1

+0x00001180    5 170          sym.fn1

 0x00001060    1 6            sym.imp.malloc

 0x00001030    1 6            sym.imp.strncpy

-0x00001230    1 48           dbg.new_some

+0x00001230    1 48           sym.new_some

 0x00001050    1 6            sym.imp.memset

-0x00001260    6 335          dbg.main

+0x00001260    6 335          main

 0x00001000    3 23           sym._init

 0x00001040    1 6            sym.imp.printf

 0x00001070    1 6            sym.imp.strdup

-            ; CALL XREF from dbg.main @ 0x132a

-            ;-- fn1:

-/ _Bool fn1(int a, char *g, double q, struct Some *gg, some_t **out)

-|           ; var char **arg1 @ rdi

-|           ; var const char *arg2 @ rsi

-|           ; var int64_t arg7 @ xmm0

-|           ; var uintmax_t arg3 @ rdx

-|           ; var some_t *arg4 @ rcx

-|           ; arg int a @ stack - 0x10

-|           ; arg char *g @ stack - 0x18

-|           ; arg double q @ stack - 0x20

-|           ; arg struct Some *gg @ stack - 0x28

-|           ; arg some_t **out @ stack - 0x30

-|           0x00001180      push  rbp                                  ; float_ex1.c:15 ; _Bool fn1(int a, char *g, double q, struct Some *gg, some_t **out)

-|           0x00001181      mov   rbp, rsp

-|           0x00001184      sub   rsp, 0x30

-|           0x00001188      mov   dword [a], edi                       ; arg1

-|           0x0000118b      mov   qword [g], rsi                       ; arg2

-|           0x0000118f      movsd qword [q], xmm0                      ; arg7

-|           0x00001194      mov   qword [gg], rdx                      ; arg3

-|           0x00001198      mov   qword [out], rcx                     ; arg4

-|           0x0000119c      cmp   qword [gg], 0                        ; float_ex1.c:16

-|       ,=< 0x000011a1      je    0x11b2

-|       |   0x000011a7      cmp   qword [out], 0

-|      ,==< 0x000011ac      jne   0x11bb

-|      |`-> 0x000011b2      mov   byte [a + 0x7], 0                    ; float_ex1.c:17

-|      |,=< 0x000011b6      jmp   0x121c

-|      `--> 0x000011bb      mov   edi, 0x18                            ; float_ex1.c:19 ; size_t size

-|       |   0x000011c0      call  sym.imp.malloc                       ; void *malloc(size_t size)

-|       |   0x000011c5      mov   rcx, rax

-|       |   0x000011c8      mov   rax, qword [out]

-|       |   0x000011cc      mov   qword [rax], rcx

-|       |   0x000011cf      mov   ecx, dword [a]                       ; float_ex1.c:20

-|       |   0x000011d2      mov   rax, qword [out]

-|       |   0x000011d6      mov   rax, qword [rax]

-|       |   0x000011d9      mov   dword [rax], ecx

-|       |   0x000011db      mov   rax, qword [out]                     ; float_ex1.c:21

-|       |   0x000011df      mov   rdi, qword [rax]

-|       |   0x000011e2      add   rdi, 4                               ; char *dest

-|       |   0x000011e6      mov   rsi, qword [g]                       ; const char *src

-|       |   0x000011ea      mov   edx, 5                               ; size_t  n

-|       |   0x000011ef      call  sym.imp.strncpy                      ; char *strncpy(char *dest, const char *src, size_t  n)

-|       |   0x000011f4      movsd xmm0, qword [q]                      ; float_ex1.c:22

-|       |   0x000011f9      cvtsd2ss xmm0, xmm0

-|       |   0x000011fd      mov   rax, qword [out]

-|       |   0x00001201      mov   rax, qword [rax]

-|       |   0x00001204      movss dword [rax + 0xc], xmm0

-|       |   0x00001209      mov   rcx, qword [gg]                      ; float_ex1.c:23

-|       |   0x0000120d      mov   rax, qword [out]

-|       |   0x00001211      mov   rax, qword [rax]

-|       |   0x00001214      mov   qword [rax + 0x10], rcx

-|       |   0x00001218      mov   byte [a + 0x7], 1                    ; float_ex1.c:24

-|       |   ; CODE XREF from dbg.fn1 @ 0x11b6

-|       `-> 0x0000121c      mov   al, byte [a + 0x7]                   ; float_ex1.c:25

-|           0x0000121f      and   al, 1

-|           0x00001221      movzx eax, al

-|           0x00001224      add   rsp, 0x30

-|           0x00001228      pop   rbp

-\           0x00001229      ret

-- stderr

exec: No such file or directory

Could not resolve address '127.0.0.1' or failed to connect

Cannot connect to 127.0.0.1:8002

No HTTP response

ERROR: Couldn't open bin for file 'http://127.0.0.1:8002/buildid/174fbe7291d83838e1d20cd0fc894a920b53b6f3/debuginfo'

[x] Analyze all flags starting with sym. and entry0 (aa)

[x] Analyze function calls

[x] Analyze len bytes of instructions for references

[x] Check for classes

[x] Analyze local variables and arguments

[x] Type matching analysis for all functions

[x] Applied 0 FLIRT signatures via sigdb

[x] Propagate noreturn information

[x] Integrate dwarf function information.

[x] Resolve pointers to data sections

[x] Use -AA or aaaa to perform additional experimental analysis.

ERROR: core: Invalid address (dbg.fn1)

ERROR: core: Error while executing command: pdf @ dbg.fn1
```

Or just run it on `linux-x64` not `linux-any`.

**Review by XVilka — CHANGES_REQUESTED** [review]
One failing CI job left (also need to install debuginfod):
```
[XX] /home/runner/work/rizin/rizin/rizin/test/db/archos/linux-x64/debuginfod "Debugging Information in debuginfod"
RZ_NOPLUGINS=1 /usr/bin/rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -eflirt.sigdb.load.system=false -eflirt.sigdb.load.home=false -N -Qc '& !debuginfod -vvv -F bins/elf/float_ex1/float_ex1_buildid bins/elf/float_ex1/float_ex1_split-dwarf.debug
sleep 10
e bin.dbginfo.debuginfod=true
e bin.dbginfo.debuginfod_urls=http://127.0.0.1:8002/
o bins/elf/float_ex1/float_ex1_buildid
aaa
iS
afl
pdf @ dbg.fn1
qyn
' =
-- stdout
--- expected
+++ actual
@@ -39,71 +39,12 @@
 0x00001080    1 6            sym.imp.__cxa_finalize
 0x00001170    1 9            entry.init0
 0x000013b0    1 9            sym._fini
-0x00001180    5 170          dbg.fn1
+0x00001180    5 170          sym.fn1
 0x00001060    1 6            sym.imp.malloc
 0x00001030    1 6            sym.imp.strncpy
-0x00001230    1 48           dbg.new_some
+0x00001230    1 48           sym.new_some
 0x00001050    1 6            sym.imp.memset
-0x00001260    6 335          dbg.main
+0x00001260    6 335          main
 0x00001000    3 23           sym._init
 0x00001040    1 6            sym.imp.printf
 0x00001070    1 6            sym.imp.strdup
-            ; CALL XREF from dbg.main @ 0x132a
-            ;-- fn1:
-/ _Bool fn1(int a, char *g, double q, struct Some *gg, some_t **out)
-|           ; var char **arg1 @ rdi
-|           ; var const char *arg2 @ rsi
-|           ; var int64_t arg7 @ xmm0
-|           ; var uintmax_t arg3 @ rdx
-|           ; var some_t *arg4 @ rcx
-|           ; arg int a @ stack - 0x10
-|           ; arg char *g @ stack - 0x18
-|           ; arg double q @ stack - 0x20
-|           ; arg struct Some *gg @ stack - 0x28
-|           ; arg some_t **out @ stack - 0x30
-|           0x00001180      push  rbp                                  ; float_ex1.c:15 ; _Bool fn1(int a, char *g, double q, struct Some *gg, some_t **out)
-|           0x00001181      mov   rbp, rsp
-|           0x00001184      sub   rsp, 0x30
-|           0x00001188      mov   dword [a], edi                       ; arg1
-|           0x0000118b      mov   qword [g], rsi                       ; arg2
-|           0x0000118f      movsd qword [q], xmm0                      ; arg7
-|           0x00001194      mov   qword [gg], rdx                      ; arg3
-|           0x00001198      mov   qword [out], rcx                     ; arg4
-|           0x0000119c      cmp   qword [gg], 0                        ; float_ex1.c:16
-|       ,=< 0x000011a1      je    0x11b2
-|       |   0x000011a7      cmp   qword [out], 0
-|      ,==< 0x000011ac      jne   0x11bb
-|      |`-> 0x000011b2      mov   byte [a + 0x7], 0                    ; float_ex1.c:17
-|      |,=< 0x000011b6      jmp   0x121c
-|      `--> 0x000011bb      mov   edi, 0x18                            ; float_ex1.c:19 ; size_t size
-|       |   0x000011c0      call  sym.imp.malloc                       ; void *malloc(size_t size)
-|       |   0x000011c5      mov   rcx, rax
-|       |   0x000011c8      mov   rax, qword [out]
-|       |   0x000011cc      mov   qword [rax], rcx
-|       |   0x000011cf      mov   ecx, dword [a]                       ; float_ex1.c:20
-|       |   0x000011d2      mov   rax, qword [out]
-|       |   0x000011d6      mov   rax, qword [rax]
-|       |   0x000011d9      mov   dword [rax], ecx
-|       |   0x000011db      mov   rax, qword [out]                     ; float_ex1.c:21
-|       |   0x000011df      mov   rdi, qword [rax]
-|       |   0x000011e2      add   rdi, 4                               ; char *dest
-|       |   0x000011e6      mov   rsi, qword [g]                       ; const char *src
-|       |   0x000011ea      mov   edx, 5                               ; size_t  n
-|       |   0x000011ef      call  sym.imp.strncpy                      ; char *strncpy(char *dest, const char *src, size_t  n)
-|       |   0x000011f4      movsd xmm0, qword [q]                      ; float_ex1.c:22
-|       |   0x000011f9      cvtsd2ss xmm0, xmm0
-|       |   0x000011fd      mov   rax, qword [out]
-|       |   0x00001201      mov   rax, qword [rax]
-|       |   0x00001204      movss dword [rax + 0xc], xmm0
-|       |   0x00001209      mov   rcx, qword [gg]                      ; float_ex1.c:23
-|       |   0x0000120d      mov   rax, qword [out]
-|       |   0x00001211      mov   rax, qword [rax]
-|       |   0x00001214      mov   qword [rax + 0x10], rcx
-|       |   0x00001218      mov   byte [a + 0x7], 1                    ; float_ex1.c:24
-|       |   ; CODE XREF from dbg.fn1 @ 0x11b6
-|       `-> 0x0000121c      mov   al, byte [a + 0x7]                   ; float_ex1.c:25
-|           0x0000121f      and   al, 1
-|           0x00001221      movzx eax, al
-|           0x00001224      add   rsp, 0x30
-|           0x00001228      pop   rbp
-\           0x00001229      ret

-- stderr
exec: No such file or directory
Could not resolve address '127.0.0.1' or failed to connect
Cannot connect to 127.0.0.1:8002
No HTTP response
ERROR: Couldn't open bin for file 'http://127.0.0.1:8002/buildid/174fbe7291d83838e1d20cd0fc894a920b53b6f3/debuginfo'
[ ] Analyze all flags starting with sym. and entry0 (aa)
[
[x] Analyze all flags starting with sym. and entry0 (aa)

[ ] Analyze function calls
[
[x] Analyze function calls
[ ] Analyze len bytes of instructions for references
[
[x] Analyze len bytes of instructions for references
[ ] Check for classes
[
[x] Check for classes
[ ] Analyze local variables and arguments
[
[x] Analyze local variables and arguments
[ ] Type matching analysis for all functions
[
[x] Type matching analysis for all functions
[ ] Applying signatures from sigdb
[
[x] Applied 0 FLIRT signatures via sigdb
[ ] Propagate noreturn information
[
[x] Propagate noreturn information
[ ] Integrate dwarf function information.
[
[x] Integrate dwarf function information.
[ ] Resolve pointers to data sections
[
[x] Resolve pointers to data sections

[x] Use -AA or aaaa to perform additional experimental analysis.
ERROR: core: Invalid address (dbg.fn1)
ERROR: core: Error while executing command: pdf @ dbg.fn1
```
See https://github.com/rizinorg/rizin/actions/runs/6760979111/job/18375352443?pr=3954#step:7:66

**Review by XVilka — APPROVED** [review]

**Comment by XVilka** [comment]
@imbillow please update the test (we removed this duplicated comment):
```
[XX] db/archos/linux-x64/debuginfod "Debugging Information in debuginfod"
RZ_NOPLUGINS=1 /usr/bin/rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -eflirt.sigdb.load.system=false -eflirt.sigdb.load.home=false -N -Qc '& !debuginfod -vvv -F bins/elf/float_ex1/float_ex1_buildid bins/elf/float_ex1/float_ex1_split-dwarf.debug
sleep 10
e bin.dbginfo.debuginfod=true
e bin.dbginfo.debuginfod_urls=http://127.0.0.1:8002/
o bins/elf/float_ex1/float_ex1_buildid
aaa
iS
afl
pdf @ dbg.fn1
qyn
' =
-- stdout
--- expected
+++ actual
@@ -49,7 +49,6 @@
 0x00001040    1 6            sym.imp.printf
 0x00001070    1 6            sym.imp.strdup
             ; CALL XREF from dbg.main @ 0x132a
-            ;-- fn1:
 / _Bool fn1(int a, char *g, double q, struct Some *gg, some_t **out)
 |           ; var char **arg1 @ rdi
 |           ; var const char *arg2 @ rsi
@@ -61,7 +60,7 @@
 |           ; arg double q @ stack - 0x20
 |           ; arg struct Some *gg @ stack - 0x28
 |           ; arg some_t **out @ stack - 0x30
-|           0x00001180      push  rbp                                  ; float_ex1.c:15 ; _Bool fn1(int a, char *g, double q, struct Some *gg, some_t **out)
+|           0x00001180      push  rbp                                  ; float_ex1.c:15
 |           0x00001181      mov   rbp, rsp
 |           0x00001184      sub   rsp, 0x30
 |           0x00001188      mov   dword [a], edi                       ; arg1
```

**Pull request merged**